File: src/personal_data_warehouse_voice_memos/writeback.py
"""Write enriched titles back into the Voice Memos app.

After the warehouse transcribes and enriches a memo, this module renames the
recording inside the Voice Memos app so the library shows the AI title instead
of "New Recording 12" or a street address. Only auto-named memos are ever
touched: a title the user typed is never overwritten.

The rename itself is a proper Core Data save against the app's
CloudRecordings.db store (see store_writer.py), which records persistent
history that the voicememod daemon exports to CloudKit — so renames sync to
every device exactly like a rename made in the app.
"""

from __future__ import annotations

from dataclasses import dataclass
import json
from pathlib import Path
import re
import sqlite3
from typing import Callable, Mapping, Sequence

import requests

# Voice Memos keeps this bit set in ZFLAGS while a recording still carries the
# name the app assigned ("New Recording N", or a geocoded location name) and
# clears it when the user types a title. Verified empirically: every
# location-auto-named row carries it, every hand-renamed row lacks it.
AUTO_NAMED_FLAG = 0x1000

# Memos recorded before the flag existed (flags 0/4 era) never carry the bit,
# but the app's non-location default names are still unmistakable.
DEFAULT_TITLE_PATTERN = re.compile(r"New Recording \d+")

MAX_TITLE_LENGTH = 200

# Persistent-history transaction author recorded with every write-back save,
# so warehouse renames are attributable in the store's own audit trail.
TRANSACTION_AUTHOR = "com.zachlatta.pdw.voice-memo-writeback"


@dataclass(frozen=True)
class LocalRecordingTitle:
    unique_id: str
    recording_id: str
    title: str
    flags: int
    filename: str = ""


@dataclass(frozen=True)
class EnrichedTitles:
    by_recording_id: dict[str, str]
    by_content_sha256: dict[str, str]


@dataclass(frozen=True)
class RenamePlanItem:
    unique_id: str
    recording_id: str
    old_title: str
    new_title: str


@dataclass(frozen=True)
class WritebackSummary:
    local_recordings: int
    auto_named: int
    enriched_titles: int
    planned: int
    renamed: int
    skipped: int
    dry_run: bool


def is_auto_named(title: str, flags: int) -> bool:
    if flags & AUTO_NAMED_FLAG:
        return True
    return DEFAULT_TITLE_PATTERN.fullmatch(title or "") is not None


def sanitize_title(title: str | None) -> str | None:
    if title is None:
        return None
    cleaned = "".join(char for char in title if char >= " " or char in "\t\n")
    collapsed = " ".join(cleaned.split())
    if not collapsed:
        return None
    return collapsed[:MAX_TITLE_LENGTH]


def load_local_recording_titles(recordings_path: Path | str) -> list[LocalRecordingTitle]:
    database_path = Path(recordings_path).expanduser() / "CloudRecordings.db"
    if not database_path.exists():
        return []
    connection = sqlite3.connect(f"file:{database_path}?mode=ro", uri=True)
    try:
        rows = connection.execute(
            """
            SELECT ZUNIQUEID, ZPATH, ZENCRYPTEDTITLE, ZFLAGS
            FROM ZCLOUDRECORDING
            WHERE ZUNIQUEID IS NOT NULL AND ZPATH IS NOT NULL
            """
        ).fetchall()
    finally:
        connection.close()
    titles: list[LocalRecordingTitle] = []
    for unique_id, path, title, flags in rows:
        filename = str(path or "")
        if not filename:
            continue
        titles.append(
            LocalRecordingTitle(
                unique_id=str(unique_id),
                recording_id=filename.rsplit(".", 1)[0],
                title=str(title or ""),
                flags=int(flags or 0),
                filename=filename,
            )
        )
    return titles


def resolve_effective_titles(
    local: Sequence[LocalRecordingTitle],
    titles: EnrichedTitles,
    sha_by_filename: Mapping[str, str] | None,
) -> dict[str, str]:
    """Map each local recording's stem to its enriched title.

    Stems normally match `enrichments.recording_id` directly, but Voice Memos
    rebases filename timestamps when the timezone changes, leaving the
    warehouse knowing a memo only under an older stem. The audio content sha
    (cached per filename in the uploader's state) still identifies it, so it
    serves as the fallback join key.
    """
    effective: dict[str, str] = {}
    for recording in local:
        title = titles.by_recording_id.get(recording.recording_id)
        if title is None and sha_by_filename:
            sha = sha_by_filename.get(recording.filename)
            if sha:
                title = titles.by_content_sha256.get(sha)
        if title is not None:
            effective[recording.recording_id] = title
    return effective


def build_rename_plan(
    local: Sequence[LocalRecordingTitle],
    enriched_titles: Mapping[str, str],
    *,
    limit: int | None = None,
) -> list[RenamePlanItem]:
    plan: list[RenamePlanItem] = []
    # Recording ids start with the recording timestamp, so sorting by id
    # descending renames the newest memos first when a limit applies.
    for recording in sorted(local, key=lambda item: item.recording_id, reverse=True):
        if not is_auto_named(recording.title, recording.flags):
            continue
        new_title = sanitize_title(enriched_titles.get(recording.recording_id))
        if new_title is None or new_title == recording.title:
            continue
        plan.append(
            RenamePlanItem(
                unique_id=recording.unique_id,
                recording_id=recording.recording_id,
                old_title=recording.title,
                new_title=new_title,
            )
        )
        if limit is not None and len(plan) >= limit:
            break
    return plan


def fetch_enriched_titles(
    *,
    base_url: str,
    secret_token: str,
    account: str,
    client_name: str = "pdw",
    session: requests.Session | None = None,
    timeout: float = 30.0,
) -> EnrichedTitles:
    """Fetch the newest completed enrichment title per recording from the app.

    Uses the same static-bearer HTTP tool API the pdw CLI's `sql` command
    uses, so no new server surface is involved. Titles come back keyed by
    recording_id and by audio content sha (the drift-proof identity).
    """
    escaped_account = account.replace("'", "''")
    sql = (
        "SELECT DISTINCT ON (recording_id) recording_id, content_sha256, title "
        "FROM apple_voice_memos.enrichments "
        "WHERE status = 'completed' AND title IS NOT NULL "
        f"AND account = '{escaped_account}' "
        "ORDER BY recording_id, created_at DESC"
    )
    http = session or requests.Session()
    response = http.post(
        f"{base_url.rstrip('/')}/api/tools/sql",
        json={
            "question": "Voice memo enriched titles for app write-back",
            "sql": sql,
            # The server's name for newline-delimited JSON (unknown formats
            # silently fall back to csv, so this string matters).
            "format": "ndjson",
        },
        headers={"Authorization": f"Bearer {client_name}:{secret_token}"},
        timeout=timeout,
    )
    response.raise_for_status()
    data = response.json().get("data") or {}
    error = data.get("error")
    if error:
        raise RuntimeError(f"enriched title query failed: {error}")
    by_recording_id: dict[str, str] = {}
    by_content_sha256: dict[str, str] = {}
    for line in (data.get("rows") or "").splitlines():
        line = line.strip()
        if not line:
            continue
        row = json.loads(line)
        recording_id = row.get("recording_id")
        title = row.get("title")
        if not recording_id or not title:
            continue
        by_recording_id[str(recording_id)] = str(title)
        sha = row.get("content_sha256")
        if sha:
            by_content_sha256.setdefault(str(sha), str(title))
    return EnrichedTitles(by_recording_id=by_recording_id, by_content_sha256=by_content_sha256)


def _default_writer(store_path, items, *, author, dry_run=False):
    from personal_data_warehouse_voice_memos.store_writer import apply_renames

    return apply_renames(store_path, items, author=author, dry_run=dry_run)


class VoiceMemosWritebackRunner:
    def __init__(
        self,
        *,
        recordings_path: Path | str,
        account: str,
        base_url: str,
        secret_token: str,
        logger,
        client_name: str = "pdw",
        writer: Callable | None = None,
        session: requests.Session | None = None,
        limit: int | None = None,
        dry_run: bool = False,
        sha_by_filename: Mapping[str, str] | None = None,
    ) -> None:
        self._recordings_path = Path(recordings_path).expanduser()
        self._account = account
        self._base_url = base_url
        self._secret_token = secret_token
        self._client_name = client_name
        self._logger = logger
        self._writer = writer or _default_writer
        self._session = session
        self._limit = limit
        self._dry_run = dry_run
        self._sha_by_filename = sha_by_filename

    def run(self) -> WritebackSummary:
        local = load_local_recording_titles(self._recordings_path)
        auto_named = [item for item in local if is_auto_named(item.title, item.flags)]
        self._logger.info(
            "Voice Memos write-back: %s local recordings, %s still auto-named",
            len(local),
            len(auto_named),
        )
        if not auto_named:
            return self._summary(local=local, auto_named=auto_named, enriched_titles={}, plan=[], results=[])

        titles = fetch_enriched_titles(
            base_url=self._base_url,
            secret_token=self._secret_token,
            account=self._account,
            client_name=self._client_name,
            session=self._session,
        )
        enriched_titles = resolve_effective_titles(local, titles, self._sha_by_filename)
        plan = build_rename_plan(local, enriched_titles, limit=self._limit)
        self._logger.info(
            "Voice Memos write-back: %s enriched titles available (%s matched locally), %s renames planned",
            len(titles.by_recording_id),
            len(enriched_titles),
            len(plan),
        )
        for item in plan:
            self._logger.info(
                "%s rename %s: %r -> %r",
                "[dry-run] would" if self._dry_run else "will",
                item.recording_id,
                item.old_title,
                item.new_title,
            )
        if self._dry_run or not plan:
            return self._summary(local=local, auto_named=auto_named, enriched_titles=enriched_titles, plan=plan, results=[])

        results = self._writer(
            self._recordings_path / "CloudRecordings.db",
            plan,
            author=TRANSACTION_AUTHOR,
        )
        for result in results:
            if result["status"] != "renamed":
                self._logger.warning(
                    "Voice Memos write-back skipped %s: %s",
                    result["unique_id"],
                    result["status"],
                )
        return self._summary(local=local, auto_named=auto_named, enriched_titles=enriched_titles, plan=plan, results=results)

    def _summary(self, *, local, auto_named, enriched_titles, plan, results) -> WritebackSummary:
        renamed = sum(1 for result in results if result["status"] == "renamed")
        return WritebackSummary(
            local_recordings=len(local),
            auto_named=len(auto_named),
            enriched_titles=len(enriched_titles),
            planned=len(plan),
            renamed=renamed,
            skipped=len(results) - renamed,
            dry_run=self._dry_run,
        )

File: app/internal/uploaders/voicememos/writeback.go
package voicememos

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/zachlatta/personal-data-warehouse/app/internal/uploaders/common"
	"github.com/zachlatta/personal-data-warehouse/app/internal/warehouse"
)

// After the warehouse transcribes and enriches a memo, the write-back renames
// the recording inside the Voice Memos app so the library shows the AI title
// instead of "New Recording 12" or a street address. Only auto-named memos
// are ever touched: a title the user typed is never overwritten. The rename
// is a real Core Data save (see storewriter.go) so it syncs to every device.

// AutoNamedFlag is the ZFLAGS bit Voice Memos keeps set while a recording
// still carries the name the app assigned, and clears when the user types a
// title.
const AutoNamedFlag = 0x1000

// MaxTitleLength caps a written title.
const MaxTitleLength = 200

// TransactionAuthor is recorded with every write-back save, so warehouse
// renames are attributable in the store's own persistent history.
const TransactionAuthor = "com.zachlatta.pdw.voice-memo-writeback"

// defaultTitlePattern matches the app's non-location default names, which
// pre-flag-era memos carry without the bit.
var defaultTitlePattern = regexp.MustCompile(`^New Recording \p{Nd}+$`)

// LocalRecordingTitle is one CloudRecordings.db row's naming state.
type LocalRecordingTitle struct {
	UniqueID    string
	RecordingID string
	Title       string
	Flags       int64
	Filename    string
}

// EnrichedTitles is the warehouse's newest completed title per recording,
// keyed both ways.
type EnrichedTitles struct {
	ByRecordingID   map[string]string
	ByContentSHA256 map[string]string
}

// RenamePlanItem is one rename to apply.
type RenamePlanItem struct {
	UniqueID    string `json:"unique_id"`
	RecordingID string `json:"recording_id"`
	OldTitle    string `json:"old_title"`
	NewTitle    string `json:"new_title"`
}

// WritebackSummary is the write-back run's counts.
type WritebackSummary struct {
	LocalRecordings int
	AutoNamed       int
	EnrichedTitles  int
	Planned         int
	Renamed         int
	Skipped         int
	DryRun          bool
}

// IsAutoNamed reports whether the app, not the user, named the recording.
func IsAutoNamed(title string, flags int64) bool {
	if flags&AutoNamedFlag != 0 {
		return true
	}
	return defaultTitlePattern.MatchString(title)
}

// SanitizeTitle drops control characters, collapses whitespace and caps the
// length; ok is false when nothing is left.
func SanitizeTitle(title string) (string, bool) {
	var cleaned strings.Builder
	for _, r := range title {
		if r >= ' ' || r == '\t' || r == '\n' {
			cleaned.WriteRune(r)
		}
	}
	collapsed := strings.Join(strings.Fields(cleaned.String()), " ")
	if collapsed == "" {
		return "", false
	}
	runes := []rune(collapsed)
	if len(runes) > MaxTitleLength {
		runes = runes[:MaxTitleLength]
	}
	return string(runes), true
}

// LoadLocalRecordingTitles reads every named recording from the store (read
// only); a missing store is an empty list.
func LoadLocalRecordingTitles(recordingsPath string) ([]LocalRecordingTitle, error) {
	databasePath := filepath.Join(common.ExpandUser(recordingsPath), "CloudRecordings.db")
	if _, err := os.Stat(databasePath); err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	db, err := common.OpenSQLiteReadOnly(databasePath)
	if err != nil {
		return nil, err
	}
	defer db.Close()
	rows, err := common.Query(db, "SELECT ZUNIQUEID, ZPATH, ZENCRYPTEDTITLE, ZFLAGS FROM ZCLOUDRECORDING WHERE ZUNIQUEID IS NOT NULL AND ZPATH IS NOT NULL")
	if err != nil {
		return nil, err
	}
	var titles []LocalRecordingTitle
	for _, row := range rows {
		filename := row.String("ZPATH")
		if filename == "" {
			continue
		}
		titles = append(titles, LocalRecordingTitle{
			UniqueID:    row.String("ZUNIQUEID"),
			RecordingID: stemOf(filename),
			Title:       row.String("ZENCRYPTEDTITLE"),
			Flags:       row.Int("ZFLAGS"),
			Filename:    filename,
		})
	}
	return titles, nil
}

// stemOf mirrors filename.rsplit(".", 1)[0].
func stemOf(filename string) string {
	if idx := strings.LastIndex(filename, "."); idx >= 0 {
		return filename[:idx]
	}
	return filename
}

// ResolveEffectiveTitles maps each local recording's stem to its enriched
// title. Stems normally match enrichments.recording_id directly, but Voice
// Memos rebases filename timestamps when the timezone changes, leaving the
// warehouse knowing a memo only under an older stem; the audio content sha
// (cached per filename in the upload state) is the fallback join key.
func ResolveEffectiveTitles(local []LocalRecordingTitle, titles EnrichedTitles, shaByFilename map[string]string) map[string]string {
	effective := map[string]string{}
	for _, recording := range local {
		title, ok := titles.ByRecordingID[recording.RecordingID]
		if !ok && len(shaByFilename) > 0 {
			if sha := shaByFilename[recording.Filename]; sha != "" {
				title, ok = titles.ByContentSHA256[sha]
			}
		}
		if ok {
			effective[recording.RecordingID] = title
		}
	}
	return effective
}

// BuildRenamePlan lists the auto-named recordings whose enriched title
// differs, newest first (recording ids start with the timestamp), stopping
// at limit when it is positive.
func BuildRenamePlan(local []LocalRecordingTitle, enrichedTitles map[string]string, limit int) []RenamePlanItem {
	sorted := append([]LocalRecordingTitle(nil), local...)
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i].RecordingID > sorted[j].RecordingID })
	var plan []RenamePlanItem
	for _, recording := range sorted {
		if !IsAutoNamed(recording.Title, recording.Flags) {
			continue
		}
		raw, ok := enrichedTitles[recording.RecordingID]
		if !ok {
			continue
		}
		newTitle, ok := SanitizeTitle(raw)
		if !ok || newTitle == recording.Title {
			continue
		}
		plan = append(plan, RenamePlanItem{UniqueID: recording.UniqueID, RecordingID: recording.RecordingID, OldTitle: recording.Title, NewTitle: newTitle})
		if limit > 0 && len(plan) >= limit {
			break
		}
	}
	return plan
}

// enrichmentsRelation is the catalog id of the enrichment attempts table.
const enrichmentsRelation = "apple_voice_memos_enrichments"

// EnrichedTitlesSQL is the statement the write-back runs through the app's
// sql tool: the newest completed enrichment title per recording for the
// account. The relation is named through the catalog, because this SQL
// crosses the HTTP tool API and is not expanded by the warehouse itself --
// and it is rendered UNQUOTED (schema.name), exactly as the Python
// fetch_enriched_titles did, so the wire statement is the same text.
func EnrichedTitlesSQL(account string) string {
	// SchemaOf panics on an id the catalog does not know, which is the
	// fail-loud check DisplayRelation alone would skip.
	_ = warehouse.SchemaOf(enrichmentsRelation)
	return "SELECT DISTINCT ON (recording_id) recording_id, content_sha256, title " +
		"FROM " + warehouse.DisplayRelation(enrichmentsRelation) + " " +
		"WHERE status = 'completed' AND title IS NOT NULL " +
		"AND account = " + common.SQLLiteral(account) + " " +
		"ORDER BY recording_id, created_at DESC"
}

// SQLQuerier runs a statement through the app's sql tool and returns rows;
// common.SQLToolQuery in production, a fake in tests.
type SQLQuerier func(question, statement string) ([]map[string]any, error)

// SQLToolQuerier binds common.SQLToolQuery to the app.
func SQLToolQuerier(baseURL, clientName, token string, timeout time.Duration) SQLQuerier {
	return func(question, statement string) ([]map[string]any, error) {
		return common.SQLToolQuery(baseURL, clientName, token, question, statement, timeout)
	}
}

// FetchEnrichedTitles asks the app for the titles, keyed by recording_id and
// by audio content sha (first title wins per sha).
func FetchEnrichedTitles(query SQLQuerier, account string) (EnrichedTitles, error) {
	rows, err := query("Voice memo enriched titles for app write-back", EnrichedTitlesSQL(account))
	if err != nil {
		return EnrichedTitles{}, fmt.Errorf("enriched title query failed: %w", err)
	}
	titles := EnrichedTitles{ByRecordingID: map[string]string{}, ByContentSHA256: map[string]string{}}
	for _, row := range rows {
		recordingID := pyStrOrEmpty(row["recording_id"])
		title := pyStrOrEmpty(row["title"])
		if recordingID == "" || title == "" {
			continue
		}
		titles.ByRecordingID[recordingID] = title
		if sha := pyStrOrEmpty(row["content_sha256"]); sha != "" {
			if _, exists := titles.ByContentSHA256[sha]; !exists {
				titles.ByContentSHA256[sha] = title
			}
		}
	}
	return titles, nil
}

func pyStrOrEmpty(value any) string {
	if value == nil {
		return ""
	}
	return common.PyStr(value)
}

// WriteResult is one plan item's outcome from the store writer.
type WriteResult struct {
	UniqueID string `json:"unique_id"`
	Status   string `json:"status"`
}

// StoreWriter applies a rename plan to CloudRecordings.db (a real Core Data
// save through the Swift helper); tests inject a fake.
type StoreWriter func(storePath string, items []RenamePlanItem, author string, dryRun bool) ([]WriteResult, error)

// WritebackRunner renames auto-named memos to their enriched titles.
type WritebackRunner struct {
	RecordingsPath string
	Account        string
	Query          SQLQuerier
	Logger         common.Logger
	Writer         StoreWriter // nil means DefaultStoreWriter
	Limit          int         // 0 means no limit
	DryRun         bool
	SHAByFilename  map[string]string
}

// Run performs one write-back pass.
func (r *WritebackRunner) Run() (WritebackSummary, error) {
	root := common.ExpandUser(r.RecordingsPath)
	local, err := LoadLocalRecordingTitles(root)
	if err != nil {
		return WritebackSummary{}, err
	}
	autoNamed := 0
	for _, item := range local {
		if IsAutoNamed(item.Title, item.Flags) {
			autoNamed++
		}
	}
	r.Logger.Infof("Voice Memos write-back: %d local recordings, %d still auto-named", len(local), autoNamed)
	summary := WritebackSummary{LocalRecordings: len(local), AutoNamed: autoNamed, DryRun: r.DryRun}
	if autoNamed == 0 {
		return summary, nil
	}
	titles, err := FetchEnrichedTitles(r.Query, r.Account)
	if err != nil {
		return summary, err
	}
	enriched := ResolveEffectiveTitles(local, titles, r.SHAByFilename)
	plan := BuildRenamePlan(local, enriched, r.Limit)
	summary.EnrichedTitles = len(enriched)
	summary.Planned = len(plan)
	r.Logger.Infof("Voice Memos write-back: %d enriched titles available (%d matched locally), %d renames planned", len(titles.ByRecordingID), len(enriched), len(plan))
	for _, item := range plan {
		verb := "will"
		if r.DryRun {
			verb = "[dry-run] would"
		}
		r.Logger.Infof("%s rename %s: %s -> %s", verb, item.RecordingID, pyRepr(item.OldTitle), pyRepr(item.NewTitle))
	}
	if r.DryRun || len(plan) == 0 {
		return summary, nil
	}
	writer := r.Writer
	if writer == nil {
		writer = DefaultStoreWriter
	}
	results, err := writer(filepath.Join(root, "CloudRecordings.db"), plan, TransactionAuthor, false)
	if err != nil {
		return summary, err
	}
	for _, result := range results {
		if result.Status == StatusRenamed {
			summary.Renamed++
		} else {
			r.Logger.Warningf("Voice Memos write-back skipped %s: %s", result.UniqueID, result.Status)
		}
	}
	summary.Skipped = len(results) - summary.Renamed
	return summary, nil
}

// pyRepr renders a string as Python's repr() does, for the log lines the
// Python write-back printed with %r.
func pyRepr(s string) string {
	quote := byte('\'')
	if strings.Contains(s, "'") && !strings.Contains(s, "\"") {
		quote = '"'
	}
	var b strings.Builder
	b.WriteByte(quote)
	for _, r := range s {
		switch {
		case r == rune(quote) || r == '\\':
			b.WriteByte('\\')
			b.WriteRune(r)
		case r == '\n':
			b.WriteString(`\n`)
		case r == '\t':
			b.WriteString(`\t`)
		case r == '\r':
			b.WriteString(`\r`)
		case r < ' ' || r == 0x7f:
			fmt.Fprintf(&b, `\x%02x`, r)
		default:
			b.WriteRune(r)
		}
	}
	b.WriteByte(quote)
	return b.String()
}
